File: dictionary(dp+dfs+bfs)/wordLadder2.java
https://www.programcreek.com/2014/06/leetcode-word-ladder-ii-java/
//https://massivealgorithms.blogspot.com/search?q=word+ladder
//find all shortest transformation sequence(s) 比如hot->hog->dog->dig和hot->dot->dog->dig
//str1 -> str2 using words in dict, one transformation can change only one letter,
To track the actual ladder, we need to add a pointer that points to the previous node in the WordNode class.
In addition, the used word can not directly removed from the dictionary. The used word is only removed when steps change.
class WordNode{
    String word;
    int numSteps;
    WordNode pre;

    public WordNode(String word, int numSteps, WordNode pre){
        this.word = word;
        this.numSteps = numSteps;
        this.pre = pre;
    }
}

public class Solution {
    public List<List<String>> findLadders(String start, String end, Set<String> dict) {
        List<List<String>> result = new ArrayList<List<String>>();
        LinkedList<WordNode> queue = new LinkedList<WordNode>();
        Set<String> visited = new HashSet<>();
        Set<String> unvisited = new HashSet<>();

        queue.add(new WordNode(start, 1, null));
        unvisited.addAll(dict);
        // ?
        dict.add(end);

        int minStep = 0;
        int preNumSteps = 0;

        while(!queue.isEmpty()){
            WordNode top = queue.remove();
            String word = top.word;
            int currNumSteps = top.numSteps;

            if(word.equals(end)){
                if(minStep == 0){
                    minStep = top.numSteps;
                }

                if(top.numSteps == minStep && minStep !=0){
                    //nothing
                    ArrayList<String> t = new ArrayList<String>();
                    t.add(top.word);
                    while(top.pre !=null){
                        t.add(0, top.pre.word);
                        top = top.pre;
                    }
                    result.add(t);
                    continue;
                }

            }

            if(preNumSteps < currNumSteps){
                unvisited.removeAll(visited);
            }

            preNumSteps = currNumSteps;

            char[] arr = word.toCharArray();
            for(int i=0; i<arr.length; i++){
                for(char c='a'; c<='z'; c++){
                    char temp = arr[i];
                    if(arr[i]!=c){
                        arr[i]=c;
                    }

                    String newWord = new String(arr);
                    if(unvisited.contains(newWord)){
                        queue.add(new WordNode(newWord, top.numSteps+1, top));
                        visited.add(newWord);
                    }

                    arr[i]=temp;
                }
            }
        }

        return result;
    }
}
//another solution:
public List<List<String>> findLadders(String start, String end,
            Set<String> dict) {
        List<List<String>> ladders = new ArrayList<List<String>>();
        Map<String, List<String>> map = new HashMap<String, List<String>>();
        Map<String, Integer> distance = new HashMap<String, Integer>();

        dict.add(start);
        dict.add(end);

        bfs(map, distance, start, end, dict);

        List<String> path = new ArrayList<String>();

        dfs(ladders, path, end, start, distance, map);

        return ladders;
    }

    void dfs(List<List<String>> ladders, List<String> path, String crt,
            String start, Map<String, Integer> distance,
            Map<String, List<String>> map) {
        path.add(crt);
        if (crt.equals(start)) {
            Collections.reverse(path);
            ladders.add(new ArrayList<String>(path));
            Collections.reverse(path);
        } else {
            for (String next : map.get(crt)) {
                if (distance.containsKey(next) && distance.get(crt) == distance.get(next) + 1) {
                    dfs(ladders, path, next, start, distance, map);
                }
            }
        }
        path.remove(path.size() - 1);
    }

    void bfs(Map<String, List<String>> map, Map<String, Integer> distance,
            String start, String end, Set<String> dict) {
        Queue<String> q = new LinkedList<String>();
        q.offer(start);
        distance.put(start, 0);
        for (String s : dict) {
            map.put(s, new ArrayList<String>());
        }

        while (!q.isEmpty()) {
            String crt = q.poll();

            List<String> nextList = expand(crt, dict);
            for (String next : nextList) {
                map.get(next).add(crt);
                if (!distance.containsKey(next)) {
                    distance.put(next, distance.get(crt) + 1);
                    q.offer(next);
                }
            }
        }
    }

    List<String> expand(String crt, Set<String> dict) {
        List<String> expansion = new ArrayList<String>();

        for (int i = 0; i < crt.length(); i++) {
            for (char ch = 'a'; ch <= 'z'; ch++) {
                if (ch != crt.charAt(i)) {
                    String expanded = crt.substring(0, i) + ch
                            + crt.substring(i + 1);
                    if (dict.contains(expanded)) {
                        expansion.add(expanded);
                    }
                }
            }
        }

        return expansion;
    }

File: dictionary(dp+dfs+bfs)/wordLadder2.py
# find all shortest transformation sequence(s) 比如hot->hog->dog->dig和hot->dot->dog->dig
'''
从 end 到 start 做一次 BFS，并且把距离 end 的距离都保存在 distance 中。
然后在从 start 到 end 做一次 DFS，每走一步必须确保离 end 的 distance 越来越近。
'''
def wordladder2(start, end, wordlist):
    wordlist.append(start)
    wordlist.append(end)
    index = build_index(wordList)
    distance = bfs(end, index)
    res = []
    return dfs(start, end, index, distance, [start], res)

def build_index(wordlist):
    index = {}
    for word in wordlist:
        for i in range(len(word)):
            base = word[:i] + "_" + word[i+1:]
            if pattern in index:
                index[pattern].add(word)
            else:
                index[pattern] = set([word])
    '''
    {'_ot': ['hot', 'dot', 'lot'],
    'h_t': ['hot', 'hit'],
    'ho_': ['hot'],
    'd_t': ['dot'],
    'do_': ['dot', 'dog'],
    '_og': ['dog', 'log', 'cog', 'cog'],  # duplicate since we have start and end to the index, so duplicate
    'd_g': ['dog'],
    'l_t': ['lot'],
    'lo_': ['lot', 'log'], ' # duplicate
    l_g': ['log'],
    'c_g': ['cog', 'cog'],   # duplicate
    'co_': ['cog', 'cog'],   # duplicate
    '_it': ['hit'],
    'hi_': ['hit']}
    '''
    return index

# return distance map
def bfs(end, index):
    distance = {end:0}  # visited
    queue = collections.deque([end])

    while queue:
        cur = queue.popleft()
        for nei in get_nei_word(cur, index):
            if nei not in distance:
                queue.append(nei)
                distance[nei] = distance[cur] + 1

    return distance

def dfs(start, end, index, distance, path, res):
    if start == end:
        res.append(path)
        return
    # branching factor:
    # depth: from start to end
    for nei in get_nei_word(start, index):
        if distance[nei] == distance[start] - 1:
            dfs(nei, end, index, distance, path+[nei], res)

def get_nei_word(word, index):
    neightbours = []
    for i in range(len(word)):
        base = word[:i] + "_" + word[i+1:]
        for nei in index.get(base, []):
            neightbours.append(nei)
    return neightbours
